### PR TITLE
neard: warn when running unsupported build on mainnet or testnet

### PR DESCRIPTION
**Pull request by mina86**
Building via `cargo build` or even `cargo build --release` produces
insufficiently-optimised build.  The recommended way to create an
executable to run in production is `make release` which enables LTO.

There is, however, no mechanism which prevents users from or warns
them about running badly-optimised nodes.

Add a mechanism which will print warnings to the logs when trying to
initialise or run a mainnet or testnet log using an executable which
was not built via `make release` or `make nightly-release`
respectively.

### CODE REVIEW

**Comment by pmnoxx** [comment]
@mina86 SGTM. I think it's a great if we can eliminate user error as one source of possible mistakes.

File: neard/src/cli.rs [hunk 1]
```diff
@@ -149,6 +148,32 @@ pub(super) struct InitCmd {
     max_gas_burnt_view: Option<Gas>,
 }
 
+fn check_release_build(chain: &str) {
+    let is_release_build =
+        option_env!("NEAR_RELEASE_BUILD").map(|value| value == "yes").unwrap_or(false);
+    let is_nightly_build = cfg!(feature = "nightly_protocol");
+    let is_mainnet_build = is_release_build && !is_nightly_build;
+    let is_testnet_build = is_release_build && is_nightly_build;
```
Comment (bowenwang1996) [inline neard/src/cli.rs:156]: No, testnet only runs stable protocols

**Review by bowenwang1996 — CHANGES_REQUESTED** [review]

File: neard/src/cli.rs [hunk 2]
```diff
@@ -149,6 +148,32 @@ pub(super) struct InitCmd {
     max_gas_burnt_view: Option<Gas>,
 }
 
+fn check_release_build(chain: &str) {
+    let is_release_build =
+        option_env!("NEAR_RELEASE_BUILD").map(|value| value == "yes").unwrap_or(false);
```
[the remark this answers is not in the document]
Comment (mina86) [reply neard/src/cli.rs:153]: RUSTFLAGS variable is read by cargo to customise options passed to the compiler but it’s cleared when rustc is actually executed.

**Comment by mina86** [comment]
@mm-near, friendly ping.

File: neard/src/cli.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmnoxx) [reply neard/src/cli.rs]: @mina86 Yes, I just checked. Unfortunately, those variables are not set. 
@matklad I wish there was a way to get `OPT_LEVEL` env variable, from the time the compilation was done.

**Review by pmnoxx — APPROVED** [review]
LGTM,

Though probably magic keyword is not needed, rather than something like

`NEAR_BUILD_TARGET=release` would be enough.

File: neard/src/cli.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mina86) [reply neard/src/cli.rs]: There shouldn’t really be any reason to have conditional compilation based on optimisation level so general sentiment is to not have that ability in Rust.

**Review by bowenwang1996 — APPROVED** [review]

**Pull request merged**